File: solutions/2026-05-09/java/2026-05-09-cyclically-rotating-a-grid.java
```java
import java.util.ArrayList;
import java.util.List;

// Problem: Cyclically Rotating a Grid
// Link: https://leetcode.com/problems/cyclically-rotating-a-grid/
//
// Approach:
// The problem involves cyclically rotating each "layer" of the grid counter-clockwise.
// An m x n grid (where m and n are even) can be visualized as concentric rectangular
// layers. We can extract each layer, perform a cyclic rotation on it, and then
// place it back into the grid.
//
// To extract a layer:
// We can iterate through the boundaries of a layer. For a layer defined by top-left
// corner (r1, c1) and bottom-right corner (r2, c2), we traverse:
// 1. Top row from left to right: grid[r1][c1] to grid[r1][c2]
// 2. Rightmost column from top to bottom: grid[r1+1][c2] to grid[r2][c2]
// 3. Bottom row from right to left: grid[r2][c2-1] to grid[r2][c1]
// 4. Leftmost column from bottom to top: grid[r2-1][c1] to grid[r1+1][c1]
//
// To perform a cyclic rotation on a layer:
// After extracting the elements of a layer into a list (or array), a cyclic rotation
// by `k` positions means the element at index `i` moves to index `(i - k) % size`
// (modulo `size` to wrap around). Since we are doing counter-clockwise rotation,
// we need to adjust the indices. If `k` is large, we can take `k % size` to get
// the effective number of rotations.
//
// To place a layer back:
// After rotating the extracted list of elements, we fill the grid cells of that layer
// in the same traversal order as extraction, but using the rotated elements.
//
// The layers are defined by the top-left corner (r1, c1) and bottom-right corner (r2, c2).
// For the outermost layer, r1=0, c1=0, r2=m-1, c2=n-1.
// For the next layer, r1=1, c1=1, r2=m-2, c2=n-2, and so on.
// This continues until r1 >= r2 or c1 >= c2.
//
// The maximum number of layers is min(m, n) / 2.
//
// Time Complexity:
// Let m be the number of rows and n be the number of columns.
// The number of elements in a layer is approximately 2*(width + height).
// For a layer, width = c2 - c1 + 1 and height = r2 - r1 + 1.
// The total number of elements in all layers is O(m*n).
// For each layer, we extract elements (O(layer_size)), rotate the list (O(layer_size)),
// and place elements back (O(layer_size)).
// The sum of layer sizes is O(m*n).
// The effective number of rotations `k` is taken modulo the size of the layer.
// Thus, the overall time complexity is O(m*n).
//
// Space Complexity:
// We use a temporary list to store the elements of each layer. The maximum size of this list
// is the size of the largest layer, which is O(m+n).
// Thus, the space complexity is O(m+n).
class Solution {
    public int[][] cyclicallyRotatingGrid(int[][] grid, int k) {
        int m = grid.length;
        int n = grid[0].length;

        // We process the grid layer by layer.
        // Each layer is defined by its top-left (r1, c1) and bottom-right (r2, c2) coordinates.
        // Layers are processed from the outermost to the innermost.
        for (int r1 = 0, c1 = 0, r2 = m - 1, c2 = n - 1; r1 < r2 && c1 < c2; r1++, c1++, r2--, c2--) {
            // Extract elements of the current layer into a list.
            List<Integer> layerElements = new ArrayList<>();

            // Traverse the top row from left to right
            for (int c = c1; c <= c2; c++) {
                layerElements.add(grid[r1][c]);
            }
            // Traverse the rightmost column from top to bottom (excluding the top-right corner already added)
            for (int r = r1 + 1; r <= r2; r++) {
                layerElements.add(grid[r][c2]);
            }
            // Traverse the bottom row from right to left (excluding the bottom-right corner already added)
            for (int c = c2 - 1; c >= c1; c--) {
                layerElements.add(grid[r2][c]);
            }
            // Traverse the leftmost column from bottom to top (excluding the bottom-left corner already added)
            for (int r = r2 - 1; r > r1; r--) {
                layerElements.add(grid[r][c1]);
            }

            // Calculate the effective number of rotations.
            // The size of the layer is the number of elements extracted.
            int layerSize = layerElements.size();
            // We only need to rotate by k % layerSize.
            // A counter-clockwise rotation by k positions means element at index i moves to (i - k) % layerSize.
            // To simplify calculations for rotation, we can think of it as shifting elements.
            // If we rotate `k` times counter-clockwise, the element that was at index `i`
            // will now be at `(i - k) % layerSize`.
            // However, when we are filling the grid, we want to know which element from the original
            // `layerElements` list should go into a specific position.
            // For a position that corresponds to the `j`-th element in the traversal order of the layer,
            // it should receive the element from `layerElements` that was `k` positions "behind" it
            // in the original sequence. This is equivalent to `layerElements.get((j - k % layerSize + layerSize) % layerSize)`.
            // A simpler way is to calculate the new position. If an element is at index `idx` in the list,
            // after `k` counter-clockwise rotations, its new index will be `(idx - k % layerSize + layerSize) % layerSize`.
            // When we are populating, for the `idx`-th position in the rotated list, it should get the element
            // that was originally at `(idx + k % layerSize) % layerSize`.
            // The modulo arithmetic needs to be handled carefully for negative results.
            int effectiveK = k % layerSize;

            // Place the rotated elements back into the grid.
            int elementIndex = 0; // Index into the layerElements list

            // Traverse the top row from left to right
            for (int c = c1; c <= c2; c++) {
                // The element at `grid[r1][c]` should be the element that was `effectiveK` positions ahead
                // in the original `layerElements` list.
                grid[r1][c] = layerElements.get((elementIndex + effectiveK) % layerSize);
                elementIndex++;
            }
            // Traverse the rightmost column from top to bottom
            for (int r = r1 + 1; r <= r2; r++) {
                grid[r][c2] = layerElements.get((elementIndex + effectiveK) % layerSize);
                elementIndex++;
            }
            // Traverse the bottom row from right to left
            for (int c = c2 - 1; c >= c1; c--) {
                grid[r2][c] = layerElements.get((elementIndex + effectiveK) % layerSize);
                elementIndex++;
            }
            // Traverse the leftmost column from bottom to top
            for (int r = r2 - 1; r > r1; r--) {
                grid[r][c1] = layerElements.get((elementIndex + effectiveK) % layerSize);
                elementIndex++;
            }
        }

        return grid;
    }
}
```

File: solutions/2026-05-09/javascript/2026-05-09-cyclically-rotating-a-grid.js
/**
 * @param {number[][]} grid
 * @param {number} k
 * @return {number[][]}
 *
 * Problem Summary: Rotates each layer of a grid cyclically counter-clockwise k times.
 * Link: https://leetcode.com/problems/cyclically-rotating-a-grid/
 *
 * Approach:
 * The problem can be solved by treating each layer of the grid as a separate list and performing a cyclic rotation on that list.
 *
 * 1. Identify Layers: We can iterate through the grid from the outermost layer inwards. A layer is defined by its top-left corner (row, col) and its bottom-right corner (m-1-row, n-1-col).
 * 2. Extract Layer Elements: For each layer, we extract its elements into a 1D array in counter-clockwise order.
 * 3. Cyclic Rotation: We then perform a cyclic rotation on this 1D array by `k` positions. Since rotations are cyclical, we can take `k` modulo the length of the 1D array to find the effective number of rotations.
 * 4. Populate Layer: After rotating the 1D array, we populate the corresponding layer in the original grid with the rotated elements, again in counter-clockwise order.
 * 5. Repeat for All Layers: We repeat this process for all layers, moving inwards. The number of layers is `min(m, n) / 2`.
 *
 * Time Complexity:
 * Let m be the number of rows and n be the number of columns.
 * Extracting elements from a layer: The perimeter of a layer at depth `d` is `2 * (m - 2*d) + 2 * (n - 2*d) - 4` (subtract 4 for the corners counted twice). This is O(m + n).
 * Rotating a 1D array of length `L`: O(L).
 * Populating a layer: O(m + n).
 * There are `min(m, n) / 2` layers.
 * In the worst case (outermost layer), the number of elements is O(m + n).
 * The total time complexity is approximately O(min(m, n) * (m + n)).
 *
 * Space Complexity:
 * We use a temporary 1D array to store the elements of each layer. The maximum size of this array is O(m + n) (for the outermost layer).
 * Therefore, the space complexity is O(m + n).
 */
var rotateGrid = function(grid, k) {
    const m = grid.length;
    const n = grid[0].length;
    const numLayers = Math.min(m, n) / 2; // Number of layers is half of the smaller dimension

    // Create a deep copy of the grid to store the rotated result
    const rotatedGrid = grid.map(row => [...row]);

    // Iterate through each layer
    for (let layer = 0; layer < numLayers; layer++) {
        // Define the boundaries of the current layer
        let top = layer;
        let bottom = m - 1 - layer;
        let left = layer;
        let right = n - 1 - layer;

        // Extract elements of the current layer into a 1D array (counter-clockwise)
        const layerElements = [];

        // Top row (left to right)
        for (let j = left; j <= right; j++) {
            layerElements.push(grid[top][j]);
        }
        // Right column (top to bottom, excluding top-right corner already added)
        for (let i = top + 1; i <= bottom; i++) {
            layerElements.push(grid[i][right]);
        }
        // Bottom row (right to left, excluding bottom-right corner already added)
        for (let j = right - 1; j >= left; j--) {
            layerElements.push(grid[bottom][j]);
        }
        // Left column (bottom to top, excluding bottom-left and top-left corners already added)
        for (let i = bottom - 1; i > top; i--) {
            layerElements.push(grid[i][left]);
        }

        // Calculate the effective number of rotations
        const effectiveK = k % layerElements.length;

        // Rotate the layer elements cyclically
        const rotatedLayerElements = new Array(layerElements.length);
        for (let i = 0; i < layerElements.length; i++) {
            rotatedLayerElements[(i + effectiveK) % layerElements.length] = layerElements[i];
        }

        // Populate the rotated elements back into the grid layer
        let elementIndex = 0;

        // Top row (left to right)
        for (let j = left; j <= right; j++) {
            rotatedGrid[top][j] = rotatedLayerElements[elementIndex++];
        }
        // Right column (top to bottom)
        for (let i = top + 1; i <= bottom; i++) {
            rotatedGrid[i][right] = rotatedLayerElements[elementIndex++];
        }
        // Bottom row (right to left)
        for (let j = right - 1; j >= left; j--) {
            rotatedGrid[bottom][j] = rotatedLayerElements[elementIndex++];
        }
        // Left column (bottom to top)
        for (let i = bottom - 1; i > top; i--) {
            rotatedGrid[i][left] = rotatedLayerElements[elementIndex++];
        }
    }

    return rotatedGrid;
};
```